src/main/java/com/fasterxml/jackson/core/json/ByteSourceJsonBootstrapper.java

```diff
@@ -230,6 +232,13 @@ public Reader constructReader() throws IOException
                 InputStream in = _in;
 
                 if (in == null) {
+                    int size = _inputEnd - _inputPtr;
+                    if (size >= 0 && size <= 8192) {
+                        // [jackson-core#488] Avoid overhead of heap ByteBuffer allocated by InputStreamReader
+                        // when processing small inputs up to 8KiB.
+                        Charset charset = Charset.forName(enc.getJavaName());
+                        return new CharBufferReader(charset.decode(ByteBuffer.wrap(_inputBuffer, _inputPtr, _inputEnd)));
```

Comment:
I wonder how this compares with constructing a string:
```java
return new StringReader(new String(_inputBuffer, _inputPtr, _inputEnd - _inputPtr, charset));
```
The string methods tend to perform very well due to extensive optimization in the jdk, however when we use a reader, the `char[]` used by `CharBufferReader` may perform better since it's not converting between bytes and chars on read.

Reply:
Good call, will benchmark and test this out. For Latin-1/ASCII compressed strings, we'll eat the cost of a byte array copy on string creation and decode should be no-op, while non-Latin-1 case may be a bit more expensive.

Reply:
The JsonEncoding class only supports Unicode charsets. UTF8, variants of UTF16, variants of UTF32.

Reply:
https://github.com/FasterXML/jackson-benchmarks/pull/9#issuecomment-1673979137 shows `new String` to `StringReader` is better, so I created https://github.com/FasterXML/jackson-core/pull/1081 as an alternative to this PR.

Reply:
> The JsonEncoding class only supports Unicode charsets. UTF8, variants of UTF16, variants of UTF32. 7 bit ASCII is a sunset of UTF8. Latin-1 is not supported.

I should have been more precise. The `Latin-1` was intended to reference [the single byte encoding compact strings JEP 254](https://openjdk.org/jeps/254), but I should have said `non-ASCII UTF-8` will be a bit more expensive in terms of memory overhead.